addOns/openapi/src/main/java/org/zaproxy/zap/extension/openapi/converter/swagger/SwaggerConverter.java

```diff
@@ -381,4 +388,137 @@ private void addOperations(OpenAPI openApi, String url, List<OperationModel> ope
                     operationHelper.getAllOperations(entry.getValue(), url + entry.getKey()));
         }
     }
+
+    /**
+     * File based parser for v2 and v3 specs that bundles external file refs.
+     *
+     * @param file V2 or V3 OpenAPI File spec, supporting external files via ref
+     * @return Populated either with a valid OpenAPI or a list of errors
+     */
+    public static SwaggerParseResult parse(File file) {
+        ParseOptions parseOptions = new ParseOptions();
+        parseOptions.setResolve(true);
+        parseOptions.setResolveFully(true);
+        SwaggerParseResult swaggerParseResult = new SwaggerParseResult();
+        List<String> errors = new ArrayList<>();
+        for (SwaggerParserExtension ex : OpenAPIV3Parser.getExtensions()) {
+            swaggerParseResult = ex.readLocation(file.getAbsolutePath(), null, parseOptions);
+            OpenAPI openAPI = swaggerParseResult.getOpenAPI();
+            if (openAPI != null && swaggerParseResult.getMessages().isEmpty()) {
+                return swaggerParseResult;
+            } else {
+                errors.addAll(swaggerParseResult.getMessages());
+            }
+        }
+        swaggerParseResult.setMessages(errors);
+        return swaggerParseResult;
+    }
+
+    public List<StructuralNodeModifier> getStructuralNodeModifiers() {
+        List<StructuralNodeModifier> structuralNodeModifiers = new ArrayList<>();
+        try {
+            OpenAPI openAPI = getOpenAPI();
+            String targetUrl;
+            if (targetUriBuilder.isEmpty()) {
+                targetUrl = null;
+            } else {
+                targetUrl = targetUriBuilder.build();
+            }
```

Comment:
This should use one of the resolved API URLs (it's incorrect when the user does not specify the target URL).